src/configure

```diff
@@ -1244,6 +1244,15 @@ elif [ "`uname`" == "Linux" ]; then
       echo MKLLIB = $MKLLIBDIR >> kaldi.mk
     fi
     echo >> kaldi.mk
+
+    echo "MKL_LD_PRELOAD := $MKLROOT/lib/intel64/libmkl_def.so \\
+                  $MKLROOT/lib/intel64/libmkl_avx2.so \\
+                  $MKLROOT/lib/intel64/libmkl_core.so \\
+                  $MKLROOT/lib/intel64/libmkl_intel_lp64.so \\
+                  $MKLROOT/lib/intel64/libmkl_intel_thread.so" >> kaldi.mk
```

Comment:
Question: when we run Kaldi binaries, we use no threading (sequential), and this calls for multithreaded BLAS.  I missed most of the fun; are we going to always use threaded BLAS backend for Python surface API? Is this something hardcoded into/required by PyTorch?

Reply:
We don't ever want to use multi-threaded BLAS, actually; it just spins its wheels for no real speed gain.   Perhaps this should be changed, but I don't know the interaction with PyTorch.

Reply:
I found that PyTorch depends on numpy, and numpy has a weird mechanism for selecting a BLAS/LAPACKE libraries at runtime. The order can be specified at build time, and MKL goes the last by default. Maybe there is an environment variable to select a preferred one, I do not know yet. I just built my own numpy from source. But as far as MKL is concerned, the default is to build against libmkl_rt.so, the Intel's suggested way of calling MKL from Python. The default is Intel threading; for sequential, the env variable `MKL_THREADING_LAYER=SEQUENTIAL` should be set. Perhaps the best solution is to add a configuration option to Kaldi's configure for that. Maybe it should be hidden, IDK, or at the least well-documented. I'm kinda reluctant to use this mechanism in general, because it adds a dependency on the environment variable. AFAICR, It's also an option to set the threading via an API call; maybe adding that call to the common initialization code for all Kaldi binaries (ParseOption is one common point) will do the trick.

In general, it's a good idea to build Kaldi libraries this way; the only thing I want to not change is how binaries work, i.e. no requirement to set an environment variable.

What do you think?

Reply:
Thanks for the background!!  I don't think this is very critical; we will probably mostly be using this branch for testing and comparison purposes, and possibly as a base for future work, rather than aiming to merge it to master in the near future; we'll be focusing on the k2 direction.